simple.zig

```diff
@@ -0,0 +1,42 @@
+const std = @import("std");
+const WordMap = std.StringHashMap(u32);
+
+pub fn main() !void {
+    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
+    const ally = &arena.allocator;
+
+    var words = WordMap.init(ally);
+    const stdin = std.io.bufferedReader(std.io.getStdIn().reader()).reader();
+    var buf: [1024]u8 = undefined;
+    while (try stdin.readUntilDelimiterOrEof(&buf, '\n')) |line| {
+        for (line) |*c| c.* = std.ascii.toLower(c.*);
+        var it = std.mem.tokenize(line, " ");
+        while (it.next()) |word| {
+            const ret = try words.getOrPut(word);
+            if (ret.found_existing) {
+                ret.entry.value += 1;
+            } else {
+                ret.entry.key = try ally.dupe(u8, word);
+                ret.entry.value = 1;
+            }
+        }
+    }
+
+    const words_slice = try ally.alloc(WordMap.Entry, words.count());
+    var i: usize = 0;
+    var it = words.iterator();
+    while (it.next()) |entry| : (i += 1) {
+        words_slice[i] = entry.*;
```

Comment:
Out of curiosity, what does `= entry.*` mean as opposed to just `= entry`?

Reply:
`foo.*` in zig is like `*foo` in C, `.*` is a pointer dereference. In this case the `next()` function returns a `?*Entry` which is unwrapped to a `*Entry` by this `while (optional_foo) |foo|` syntax. In order to store the struct into our allocated slice by value instead of by pointer we need to deference here.

In hindsight, we could store pointers in the allocated slice for sorting instead of values (which might make things faster) as we will no longer mutate the hashmap and invalidate pointers to its entries.

Reply:
In my quick test, sorting pointers ends up being 1-10% slower, I am guessing due to dereferencing cost on each compare.